docs/buffer-metatable.md

```diff
@@ -0,0 +1,114 @@
+# Extend buffer with a metatable
+
+## Summary
+
+Extend `buffer.create` with a second argument which provides a metatable for that buffer.
+The metatable and some of its keys will be frozen for potential improvements in access.
+
+## Motivation
+
+Providing a buffer object with a metatable will allow creation of objects with small storage footprint which can also be matched to native structures on the side of the application using Luau. 
+
+By having a metatable, these buffer objects can have properties, methods and other kinds of operations.
+
+This will also allow the behavior to be defined and extended on the Luau side, compared to userdata that is strictly defined by the host.
+
+In a way, this will provide an alternative to `ctypes` in luajit.
+
+```luau
+-- A float4 in a buffer
+local mt = {
+  __index = function(b, field)
+    if field == "x" then return buffer.readf32(b, 0)
+    elseif field == "y" then return buffer.readf32(b, 4)
+    elseif field == "z" then return buffer.readf32(b, 8)
+    elseif field == "w" then return buffer.readf32(b, 12)
+    else error("unknown field") end
+  end,
+  __newindex = function(b, field, value)
+    if field == "x" then buffer.writef32(b, 0, value)
+    elseif field == "y" then buffer.writef32(b, 4, value)
+    elseif field == "z" then buffer.writef32(b, 8, value)
+    elseif field == "w" then buffer.writef32(b, 12, value)
+    else error("unknown field") end
+  end
+}
+
+local buf = buffer.create(16, mt)
+
+buf.x = 2
+buf.y = 4
+
+assert(buf.x + buf.y + buf.z == 6)
+```
+
+Or alternatively:
+```luau
+local mt = {
+  __index = {
+    x = function(b) return buffer.readf32(b, 0) end,
+    y = function(b) return buffer.readf32(b, 4) end,
+    z = function(b) return buffer.readf32(b, 8) end,
+    w = function(b) return buffer.readf32(b, 12) end,
+    setx = function(b, value) buffer.writef32(b, 0, value) end,
+    sety = function(b, value) buffer.writef32(b, 4, value) end,
+    setz = function(b, value) buffer.writef32(b, 8, value) end,
+    setw = function(b, value) buffer.writef32(b, 12, value) end,
+    magnitude = function(b) return math.sqrt(b.x * b.x + b.y * b.y + b.z * b.z + b.w * b.w) end,
+    normalize = function(b) return ... end,
+  },
+}
+
+local buf = buffer.create(16, mt)
+
+buf:setx(2)
+buf:sety(4)
+buf:normalize()
+
+local xn = buf:x()
+```
+
+Or any other custom way the developer wants property access to be performed.
+
+## Design
+
+`buffer.create(size: number, metatable: table): buffer`
+
+Buffer construction will accept a second argument which will be a metatable for the buffer object.
+
+This metatable will be frozen.
+In addition to that, `__index` and `__newindex` fields will also be frozen if they are a table.
```

Comment:
Is this recursive, or does it end here?

Reply:
It ends at this level.
You can have a dynamic nested __index for fallback accesses, but optimizations will not proceed there (you can freeze nested tables manually).